src/app/api/user.service.ts

```diff
@@ -0,0 +1,89 @@
+import { Injectable } from '@angular/core';
+import { BaseApiService, Version } from './base-api.service';
+
+export interface User {
+  id: number;
+  username: string;
+}
+
+@Injectable()
+export class UserService {
+  private loggedIn = false;
+  private user: User;
+
+  constructor(private api: BaseApiService) {
+    // check if they are logged in
+    this.getDetails().then((user) => {
+      console.log(user);
+    }).catch((err) => {
+      console.error('User is not logged in');
+    });
+  }
+
+  isLoggedIn() {
+    return this.loggedIn;
+  }
+
+  isUsernameAvailable() {
+    throw new Error('Method not implemented');
+  }
+
+  /**
+   * Logs out the currently logged in user
+   */
+  logout() {
+    if (!this.isLoggedIn()) return;
+
+    return this.api.put(Version.v1, 'auth/logout', {}).then((user) => {
+      this.user = null;
+      this.loggedIn = false;
+    });
+  }
+
+  /**
+   * Logs in a new user and returns their details
+   * @param {string} username
+   * @param {string} password
+   * @return {Promise<User>}
+   */
+  login(username: string, password: string): Promise<User> {
+    return this.api.post(Version.v1, 'auth', {username, password}).then((user: User) => {
+      this.user = user;
+      this.loggedIn = true;
+      return user;
+    });
+  }
+
+  /**
+   * Signs up a new user and automatically logs them in
+   * @param {string} email
+   * @param {string} username
+   * @param {string} password
+   * @return {Promise<User>} New user details
+   */
+  signup(email: string, username: string, password: string): Promise<User> {
+    return this.api.post(Version.v1, 'users', {email, username, password}).then((user: User) => {
```

Comment:
What is the purpose of email in our situation? (I understand the purpose of an email, but we're not actually using it in our service). In other words:
- We do not ensure that the email is valid
- We do not send them an email to get registration code
- We do not ensure that emails are unique (in the backend, you ensure that it is an email, but not that it's unique i think?)
- And we never use the email for any purpose.

Reply:
The backend currently ensures the email is case insensitive unique and checks for proper structure.

We could make the email optional, since in the future if someone wants to recover their account, we'd want to let them do so.